quickjs.c

```diff
@@ -3012,19 +3012,22 @@ static const char *JS_AtomGetStrRT(JSRuntime *rt, char *buf, int buf_size,
 {
     if (__JS_AtomIsTaggedInt(atom)) {
         snprintf(buf, buf_size, "%u", __JS_AtomToUInt32(atom));
-    } else {
-        JSAtomStruct *p;
+    } else if (atom == JS_ATOM_NULL) {
+        snprintf(buf, buf_size, "<null>");
+    } else if (atom >= rt->atom_size) {
         assert(atom < rt->atom_size);
```

Comment:
does the assert make sense now?

Reply:
Yes: in the debug build the assert will crash on the spot.  In non debug builds, the atom string will show as `<invalid 1f4>` or `<free 3f8>` in the debug traces which gives the opportunity to investigate.  I used to investigate a crash in `DUMP_FREE` caused by tracing objects contents during the GC cycle removal phase (fixed in this PR).

Reply:
You have this code now:

```
} else if (atom >= rt->atom_size) {
        assert(atom < rt->atom_size);
```

So That assert can never hit because it's guared by the if that check for the opposite. Maybe you intended to place it elsewhere?

Reply:
On the contrary, the test is always false hence the `assert` is always triggered in debug mode.
Whether the compiler generates an actual test or determines its result from the context should not matter.

Reply:
Sorry maybe I'm obtuse today 😅

How will the code enter the if, if the condition is false? The assert is inside that if isn't it?

Reply:
The way I understand it, `atom >= rt->atom_size` is the error condition, once we enter in that `if` branch we know we'll fail the assert or print (when `-DNDEBUG`). We might have as well written `assert(0)` or something, but that would be less clear when inspecting the reason why it triggered.

Reply:
`atom >= rt->atom_size` is an abnormal condition.  If we are in a debug build, the `assert()` will output the `assertion failed: atom < rt->atom_size` and abort the process.  If we are not in a debug build, the atom will be converted to `<invalid ###>` or `<free ###>` without any access to the `JSAtom` structure and this will appear in the traces.

Reply:
Aaaah ok! Since I saw the condition I thought we were trying to handle it gracefully. Makes sense now!